### PR TITLE
Add NuGet.Services.AzureManagement

### PR DESCRIPTION
**Pull request by skofman1**
Uses Service Principal to get access token to use with Azure Management API.  Get cloud service properties and provides a helper method to parse the result.

### CODE REVIEW

File: src/NuGet.Services.AzureManagement/AzureHelper.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+using System;
+using System.Xml;
+using Newtonsoft.Json.Linq;
+
+namespace NuGet.Services.AzureManagement
+{
+    public static class AzureHelper
+    {
+        public class CloudService
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureHelper.cs]: Perhaps rename to `CloudServiceInfo` or `CloudServiceProperties`?

File: src/NuGet.Services.AzureManagement/AzureHelper.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            public Uri Uri { get; set; }
+            public int InstanceCount { get; set; }
+        }
+
+        /// <summary>
+        /// Parses the result of https://management.azure.com/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.ClassicCompute/domainNames/{2}/slots/{3}?api-version=2016-11-01
+        /// </summary>
+        public static CloudService ParseCloudServiceProperties(string propertiesString)
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureHelper.cs]: The method name indicates the intent is to parse cloud service properties and return this info. The return type is `CloudService`. Perhaps rename the return type to better indicate you're actually getting a hold of cloudservice info or properties, and not the actual cloud service?

File: src/NuGet.Services.AzureManagement/AzureHelper.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            try
+            {
+                var cloudService = new CloudService();
+
+                var jObject = JObject.Parse(propertiesString);
+                cloudService.Uri = new Uri(jObject["properties"]["uri"].ToString());
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureHelper.cs]: I see you use `jObject["properties"]` twice in this method. Perhaps call it once and reuse instead?

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace NuGet.Services.AzureManagement
+{
+    public class AzureManagementAPIWrapper : IAzureManagementAPIWrapper
+    {
+        private const string Authority = @"https://login.microsoftonline.com/microsoft.onmicrosoft.com";
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: nit: why the `@`?

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        private const string Resource = "https://management.core.windows.net/";
+        private const int RenewTokenPriorToExpirationMinutes = 5;
+
+        private string _clientId;
+        private string _clientSecret;
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: I think both `_clientId` and `_clientSecret` can be made `readonly`.

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                _accessToken = authenticationResult.AccessToken;
+                _tokenExpirationTime = authenticationResult.ExpiresOn;
+            }
+        }
+
+        private async Task<AuthenticationResult> GetAccessToken()
+        {
+            try
+            {
+                var clientCredential = new ClientCredential(_clientId, _clientSecret);
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: As both parameters can be readonly fields, I don't think you need to new up a new `clientCredential` on every call. It won't change during the lifetime of this instance.

File: src/NuGet.Services.AzureManagement/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+[assembly: AssemblyTitle("NuGet.Services.AzureManagement")]
+[assembly: AssemblyDescription("Azure Management tools for NuGet services")]
+[assembly: AssemblyProduct("NuGet.Services.AzureManagement")]
+[assembly: AssemblyCopyright("Copyright © .NET Foundation 2015")]
```
Comment (xavierdecoster) [inline src/NuGet.Services.AzureManagement/Properties/AssemblyInfo.cs]: Welcome to 2017! :) 🎉

File: tests/NuGet.Services.AzureManagement.Tests/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("NuGet.Services.AzureManagement.Tests")]
+[assembly: AssemblyDescription("Unit tests for NuGet.Services.AzureManagement library")]
+[assembly: AssemblyProduct("NuGet.Services.AzureManagement.Tests")]
+[assembly: AssemblyCopyright("Copyright © .NET Foundation 2016")]
```
Comment (xavierdecoster) [inline tests/NuGet.Services.AzureManagement.Tests/Properties/AssemblyInfo.cs]: 2017

File: tests/NuGet.Services.AzureManagement.Tests/project.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "dependencies": {
+    "FluentAssertions": "4.9.0",
+    "Moq": "4.6.38-alpha",
```
Comment (xavierdecoster) [inline tests/NuGet.Services.AzureManagement.Tests/project.json]: Isn't there a stable Moq version we can rely on?

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            }
+
+            if (string.IsNullOrEmpty(slot))
+            {
+                throw new ArgumentException(nameof(slot));
+            }
+
+            await RenewAccessToken();
+
+            const string RequestUrlFormat = @"https://management.azure.com/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.ClassicCompute/domainNames/{2}/slots/{3}?api-version=2016-11-01";
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: Should this be a field as well?

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+            string requestUrl = string.Format(RequestUrlFormat, subscription, resourceGroup, name, slot);
+
+            using (var client = new HttpClient())
+            {
+                var request = new HttpRequestMessage(HttpMethod.Get, requestUrl);
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
+                HttpResponseMessage response = await client.SendAsync(request, token);
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: nit: use `var` here?

File: src/NuGet.Services.AzureManagement/AzureHelper.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    public static class AzureHelper
+    {
+        public class CloudServiceProperties
+        {
+            public Uri Uri { get; set; }
+            public int InstanceCount { get; set; }
+        }
+
+        /// <summary>
+        /// Parses the result of https://management.azure.com/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.ClassicCompute/domainNames/{2}/slots/{3}?api-version=2016-11-01
```
Comment (ryuyu) [inline src/NuGet.Services.AzureManagement/AzureHelper.cs]: Is there a doc that actually details the structure of the return? Or are we assuming that this is in flux, and we will need to respond to any changes?

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+                if (response.IsSuccessStatusCode)
+                {
+                    string result = await response.Content.ReadAsStringAsync();
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: nit: return directly? Same with line 118.

File: src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace NuGet.Services.AzureManagement
+{
+    public class AzureManagementAPIWrapper : IAzureManagementAPIWrapper
+    {
+        private const string Authority = "https://login.microsoftonline.com/microsoft.onmicrosoft.com";
+        private const string Resource = "https://management.core.windows.net/";
+        private const int RenewTokenPriorToExpirationMinutes = 5;
```
Comment (shishirx34) [inline src/NuGet.Services.AzureManagement/AzureManagementAPIWrapper.cs]: why 5 minutes? why not 10?

**Pull request merged**